hathor/consensus/poa/poa_block_producer.py

```diff
@@ -113,21 +103,49 @@ def _get_signer_index(self, previous_block: Block) -> int | None:
         except ValueError:
             return None
 
-    def _start_producing(self) -> None:
+    def _safe_start_producing(self) -> None:
+        try:
+            return self._unsafe_start_producing()
+        except Exception:
+            self._log.exception('error while trying to start block production')
+
+    def _unsafe_start_producing(self) -> None:
         """Start producing new blocks."""
         if not self.manager.can_start_mining():
             # We're syncing, so we'll try again later
             self._log.warn('cannot start producing new blocks, node not synced')
             return
 
         self._log.info('started producing new blocks')
-        self._started_producing = True
-        self._start_producing_lc.stop()
+        self._schedule_block()
+
+    def _on_new_vertex(self, event: HathorEvents, args: EventArguments) -> None:
+        """Handle propagation of new blocks after a vertex is received."""
+        assert event is HathorEvents.NETWORK_NEW_TX_ACCEPTED
+        block = args.tx
+
+        from hathor.transaction import Block
+        if not isinstance(block, Block):
+            return
+
+        from hathor.transaction.poa import PoaBlock
+        if isinstance(block, PoaBlock) and not block.weight == poa.BLOCK_WEIGHT_IN_TURN:
+            self._log.info('received out of turn block', block=block.hash_hex, signer_id=block.signer_id)
+
+        self._schedule_block()
 
     def _schedule_block(self) -> None:
         """Schedule propagation of a new block."""
+        if not self.manager.can_start_mining():
+            # We're syncing, so we'll try again later
+            self._log.info('cannot produce new block, node not synced')
```

Comment:
I feel that the full node would not recover from getting out of sync after it was already synced. From recovery, I mean that this full node would stop producing blocks.

Reply:
The full node recovers as long as it receives a block from another peer. I added an assertion in an existing test to cover this, in https://github.com/HathorNetwork/hathor-core/pull/1080/commits/0d4fea63306a70d5d0d4b347fa2d0b3ef3a1b97b